staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -177,119 +202,187 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t
-	// step 4: wait for the HTTP Server listener to have stopped
-	httpServerStoppedListeningCh := s.lifecycleSignals.HTTPServerStoppedListening
-	select {
-	case <-httpServerStoppedListeningCh.Signaled():
-	case <-time.After(5 * time.Second):
-		t.Fatal("Expected the server to signal HTTPServerStoppedListening event")
+		if err := assertResponseStatusCode(resultGot, http.StatusInternalServerError); err != nil {
+			return true, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("Expected /readyz to return 500 status code, but got: %v", err)
 	}
 
-	// step 5: the server has stopped listening but we still have a request
+	// before ShutdownDelayDuration elapses new request(s) should be served successfully.
+	beforeShutdownDelayDurationStep.execute(func() {
+		t.Log("Before ShutdownDelayDuration elapses new request(s) should be served")
+		resultGot := doer.Do(connReusingClient, shouldReuseConnection(t), "/echo?message=request-on-an-existing-connection-should-succeed", time.Second)
+		if err := assertResponseStatusCode(resultGot, http.StatusOK); err != nil {
+			t.Errorf("%s", err.Error())
+		}
+		resultGot = doer.Do(newClient(true), shouldUseNewConnection(t), "/echo?message=request-on-a-new-tcp-connection-should-succeed", time.Second)
+		if err := assertResponseStatusCode(resultGot, http.StatusOK); err != nil {
+			t.Errorf("%s", err.Error())
+		}
+	})
+
+	waitForeverUntilSignaled(t, signals.AfterShutdownDelayDuration)
+
+	// preshutdown hook has not completed yet, new incomng request should succeed
+	resultGot = doer.Do(newClient(true), shouldUseNewConnection(t), "/echo?message=request-on-a-new-tcp-connection-should-succeed", time.Second)
+	if err := assertResponseStatusCode(resultGot, http.StatusOK); err != nil {
+		t.Errorf("%s", err.Error())
+	}
+
+	// let the preshutdown hook issue an API call now, and then
+	// let's wait for it to return the result.
+	close(preShutdownHook.blockedCh)
+	preShutdownHookResult := <-preShutdownHook.resultCh
+	if err := assertResponseStatusCode(preShutdownHookResult, http.StatusOK); err != nil {
+		t.Errorf("%s", err.Error())
+	}
+
+	waitForeverUntilSignaled(t, signals.PreShutdownHooksCompleted)
+	// both AfterShutdownDelayDuration and PreShutdownHooksCompleted
+	// have been signaled, we should not be accepting new request
+	waitForeverUntilSignaled(t, signals.NotAcceptingNewRequest)
+	waitForeverUntilSignaled(t, signals.HTTPServerStoppedListening)
+
+	// our original intention was for any incoming request to receive a 503
+	// via the WithWaitGroup filter, but, at this point, any incoming requests
+	// will get a 'connection refused' error since the net/http server has
+	// stopped listening.
+	// TODO (tkashem): can we consolidate these two modes of graceful termination?
+	//
+	// NOTE: the following two calls produce a flake:
+	//  Expected error connection refused, but got: <nil>
+	//  from stress tool: 10m10s: 3359 runs so far, 4 failures (0.12%)
+	//  the following sleep brings it down to zero
```

Comment:
I'm really not a fan of Sleeps - it is asking for flakes in heavily cpu-starved environment.

Why those were succeeding? Because HTTPServerStoppedListening signal is actually signalled too early (when http server is still actually listening)? If so, we should try to fix it...

Reply:
yes, I removed the sleep and put an workaround for now. It's strange really, `HTTPServerStoppedListening` is signaled when when `Server.Serve` returns. 

I don't understand exactly what's happening here, I will need to do a deep dive into the net/http code. maybe both new requests are riding on an existing connection.

https://github.com/golang/go/blob/c087121dafc1e0250d43b545e85d67e7e6762f74/src/net/http/server.go#L3063-L3065

https://github.com/golang/go/blob/c087121dafc1e0250d43b545e85d67e7e6762f74/src/net/http/server.go#L2771-L2776